### PR TITLE
Warrior Knight Strain

### PR DESCRIPTION
**Pull request by carlarctg**
## About The Pull Request

https://hackmd.io/_GMmQVHvQESXPBlDFlTPfw?view

Link to PR feedback: https://discord.com/channels/150315577943130112/1045088712712794142

TL;DR: Warrior strain meant to focus on frontline combat.

Adds a new Warrior strain, Knight.

Passive: Immortal Will

Replaces tail stab with tail trip. Tail trip acts similarly to dancer's trip, causing a flickerstun. Cannot be enhanced.

When on weeds, enhance your abilities.

Ability 1: Pike

Extend your tail 3 tiles in front of you, dealing damage across that lane. Every enemy hit in total will increase said damage by 10. When enhanced, the third tile will deal more damage and halve the cooldown on the next pike.

Ability 2: Bulwark

Gain 20% resistance to bullets which are not aimed directly towards you. Gain one stack of clarity that blocks one of any stun. If enhanced, resistance upped to 40% and gain 3 stacks of clarity instead. Ability can only be enhanced or unenhanced at activation - losing enhancement midway through, or gaining it, will not affect things.

Ability 3: Jump

Jump towards a tile, maximum 4. This ability passes through mobs. If hitting a target in an L-shape, trample them, causing stun, damage, and throw them back. When enhanced, trampling is made more powerful.

## Why It's Good For The Game

> Passive: Immortal Will
> Replaces tail stab with tail trip. Tail trip acts similarly to dancer's trip, causing a flickerstun. Cannot be enhanced.

Tail stab would be redundant with Pike, and thus was replaced with tail trip, which is a 0.5s stun meant just as an extra tool.

> When on weeds, enhance your abilities.
This will let the Knight be strong defensively while not being overpowered offensively, hopefully meaning longer rounds!!!

> Ability 1: Pike
> Extend your tail 3 tiles in front of you, dealing damage across that lane. Every enemy hit in total will increase said damage by 10. When enhanced, the third tile will deal more damage and halve the cooldown on the next pike.

The workhorse ability, will be utilized to harass the frontline at a distance that can be punished, with difficulty. The enhanced attack will be gratifying to combo with, while also being easy to dodge. Someone rushing the Knight will not be affected, and someone staying far won't either.

> Ability 2: Bulwark
> Gain 20% resistance to bullets which are not aimed directly towards you. Gain one stack of clarity that blocks one of any stun. If enhanced, resistance upped to 40% and gain 3 stacks of clarity instead. Ability can only be enhanced or unenhanced at activation - losing enhancement midway through, or gaining it, will not affect things.

Intended to fight on the frontlines without being instantly decimated by suppressive fire, of the normal or stunning variety. Numbers can be tuned in testing.

> Ability 3: Jump
> Jump towards a tile, maximum 4. This ability passes through mobs. If hitting a target in an L-shape, trample them, causing stun, damage, and throw them back. When enhanced, trampling is made more powerful.

Intended to allow the Knight to reposition around the battlefield and not bodyblock other xenomorphs, primarily T3s. The trample is intended as another gratifying trick to pull off, but is not very powerful by itself, only being as strong as a slash or two with a short stun.

This strain as a whole is intended to fit the unfilled niche of a frontline T2. Defenders can do it, T3s can do it, no T2s can, not even Spitter really.

## Changelog

:cl:
add: Added a new warrior strain. Knight. Focuses on frontline combat.
imageadd: Passive: Immortal Will
balance: Replaces tail stab with tail trip. Tail trip acts similarly to dancer's trip, causing a flickerstun. Cannot be enhanced.
balance: When on weeds, enhance your abilities.
imageadd: Ability 1: Pike
balance: Extend your tail 3 tiles in front of you, dealing damage across that lane. Every enemy hit in total will increase said damage by 10. When enhanced, the third tile will deal more damage and halve the cooldown on the next pike.
imageadd: Ability 2: Bulwark
balance: Gain 20% resistance to bullets which are not aimed directly towards you. Gain one stack of clarity that blocks one of any stun. 
balance: If Bulwark is enhanced, resistance upped to 40% and gain 3 stacks of clarity instead. Ability can only be enhanced or unenhanced at activation - losing enhancement midway through, or gaining it, will not affect things.
imageadd: Ability 3: Jump
balance: Jump towards a tile, maximum 4. This ability passes through mobs. If hitting a target in an L-shape, trample them, causing stun, damage, and throw them back. When enhanced, trampling is made more powerful.
/:cl:

### CODE REVIEW

**Comment by carlarctg** [comment]
https://streamable.com/5dp68c

**Comment by carlarctg** [comment]
I think it would be really cool if someone reviewed or testmerged this in less than a month after this comment. I think that would be really awesome!

**Review by morrowwolf — CHANGES_REQUESTED** [review]
Bugs and comments as per the PR-Feedback thread.

File: code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm [hunk 1]
```diff
@@ -195,3 +195,322 @@
...
+	RegisterSignal(Knight, COMSIG_LIVING_AMMO_KNOCKBACK, PROC_REF(resist_knockback))
+	RegisterSignal(Knight, COMSIG_ITEM_ATTEMPT_ATTACK, PROC_REF(resist_melee))
+	button.icon_state = "template_active"
+	button.color = knight_delegate.current_color
+	current_shield_image = Knight.create_bulwark_image(img_alpha_mult, "full")
+	addtimer(CALLBACK(src, PROC_REF(remove_shield)), shield_duration)
+
+	apply_cooldown()
+	..()
+	return
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm:355]: is there a reason you're not returning ..()?

File: code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm [hunk 2]
```diff
@@ -195,3 +195,322 @@
...
+
+/datum/action/xeno_action/onclick/bulwark/proc/reduce_bullet_damage(mob/xeno_byproduct_param, list/projectiledata)
+	SIGNAL_HANDLER
+	var/mob/living/carbon/xenomorph/zenomorf = owner
+
+	if(zenomorf.mutation_type != WARRIOR_KNIGHT) //why do we need these checks again? it's not like it wouldnt be a bug that needs looking into if they didnt exist. if anything this removes helpful runtimes
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm:361]: honestly i hate sanity checks

File: code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm [hunk 3]
```diff
@@ -195,3 +195,322 @@
...
+	SIGNAL_HANDLER
+	attacker.animation_attack_on(Knight)
+	attacker.flick_attack_overlay(Knight, "punch") //"shield"
+	if(hitting_item.force < MELEE_FORCE_NORMAL)
+		var/picked_attack_verb = pick(hitting_item.attack_verb)
+		picked_attack_verb = deconjugate(picked_attack_verb)
+		if(isnull(picked_attack_verb))
+			picked_attack_verb = "attack"
+		attacker.visible_message(SPAN_DANGER("[attacker] tries to [picked_attack_verb] [Knight] with \the [hitting_item], but the attack bounces right off its defensive shield!"),\
+		SPAN_DANGER("You try to [picked_attack_verb] [Knight] with \the [hitting_item], but your attack bounces right off its defensive shield!"), message_flags = CHAT_TYPE_MELEE_HIT)
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/warrior/warrior_powers.dm:418]:
```suggestion
		attacker.visible_message(SPAN_DANGER("[attacker] tries to [picked_attack_verb] [Knight] with [hitting_item], but the attack bounces right off its defensive shield!"),\
		SPAN_DANGER("You try to [picked_attack_verb] [Knight] with [hitting_item], but your attack bounces right off its defensive shield!"), message_flags = CHAT_TYPE_MELEE_HIT)
```

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/warrior/knight.dm [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+/datum/behavior_delegate/warrior_knight/proc/buff()
+
+	abilities_enhanced = TRUE
+	current_color = enhanced_color
+
+	var/color = enhanced_color
+	var/alpha = 25
+	if(COOLDOWN_FINISHED(src, toggle_msg_cd))
+		bound_xeno.visible_message(SPAN_DANGER("\The [bound_xeno] begins faintly glowing purple."), SPAN_NOTICE("Tendrils from your hive's weeds connect you to the hive's psychic power. Your abilities are now enhanced."))
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/mutators/strains/warrior/knight.dm:121]:
```suggestion
		bound_xeno.visible_message(SPAN_DANGER("[bound_xeno] begins faintly glowing purple."), SPAN_NOTICE("Tendrils from your hive's weeds connect you to the hive's psychic power. Your abilities are now enhanced."))
```

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/warrior/knight.dm [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		COOLDOWN_START(src, toggle_msg_cd, 8 SECONDS)
+	color += num2text(alpha, 2, 16)
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/mutators/strains/warrior/knight.dm:123]: shouldn't you just be assigning color here?

File: code/modules/mob/mob.dm [hunk 6]
```diff
@@ -899,6 +899,8 @@ note dizziness decrements automatically in the mob's Life() proc.
 	handle_slowed()
 	handle_superslowed()
 
+//not touching this.
+
```
Comment (harryob) [inline code/modules/mob/mob.dm:903]: i know you're not by the absence of code touching it!!!!

**Review by harryob — CHANGES_REQUESTED** [review]

**Comment by ondrej008** [comment]
I'm not sure whether we're supposed to use the forums, discord or this, so I'm doing it here.

1. Bug: The Pike ability seems to be silver colored on every tile.

2. What is the point of a frontline strain that gets stronger when fighting on weeds? I have never seen the frontline be on weeds, only time I can think of is when xenos are losing by a landslide and marines are a couple minutes from ending the round.

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: 6a2cf07b91be24fbcbc651e8a46a8ae6445dd924
Pull Request: c70fbbd74bfa6b719cb80d8c2bae3e5a37bf9e13
Server: 0f1dac4dcb22794ac0797cee8eb9c5b0b7c5e7f8

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Removed

**Pull request closed without merging**